src/tools/rustfmt/src/source_file.rs

```diff
@@ -66,17 +77,6 @@ where
         }
     }
 
-    impl From<&FileName> for rustc_span::FileName {
-        fn from(filename: &FileName) -> rustc_span::FileName {
-            match filename {
-                FileName::Real(path) => {
-                    rustc_span::FileName::Real(rustc_span::RealFileName::LocalPath(path.to_owned()))
-                }
-                FileName::Stdin => rustc_span::FileName::Custom("stdin".to_owned()),
-            }
-        }
-    }
-
```

Comment:
Can you revert this change and add `#[allow(non_local_definitions)]` here?

Reply:
Yes, the rustfmt team would prefer these types of changes be made in the rustfmt repo, so we'd rather you add an `#[allow]`.

Reply:
Sure, but what's the difference between putting a `#[allow]` and moving the code around? They both result in a subtree change.

*I'm just trying to understand the reason. It seems (in my eyes) like you don't want  any update to the subtree, which goes backwards of the idea of a subtree.*

Reply:
This particular change likely won't result in any merge conflicts when we eventually sync the subtree, but we'd like to minimize the changes made to the subtree if possible so that the sync process is smoother for us.

Reply:
@Urgau, I'm hoping we can continue the discussion. As I mentioned, it's not about preventing changes to the subtree in rust-lang/rust, it's about reducing the impact on us subtree maintainers.

Kindly linking to some recent comments from other subtree maintainers around this topic:
https://github.com/rust-lang/rust/pull/116033#issuecomment-1763251247
https://github.com/rust-lang/rust/pull/116033#issuecomment-1774119138
https://github.com/rust-lang/rust/pull/118692#issuecomment-1848629671

Reply:
I'm not dismissing your arguments. I just not happy at all about this. It's forcing every contributors to do all the heavy lifting: update upstream, minimal change downstream, ask and wait for subtree update. I don't really see what's the advantage comparing to submodules.

I've had to deal with this with r-a and while I didn't like it, the maintainer actually helped by doing the change upstream, immediately followed by a subtree update. This is not suggesting that you (or any other maintainers) do this, but this was actually somewhat helpful.

I would also note, that clippy is also a subtree where I've made more significant changes in rust-lang/rust. Don't they have the same problem?

Reply:
Thanks for the response, and taking the time to explain your perspective. Let me see if I can explain why I don't think the ask will put any undue burden on contributors. I know you're hard at work here trying to implement an accepted RFC, and in order to no block your work and to minimize the impact on rustfmt I'd like to opt-out of changes with an `#[accept]`. I think it's a win-win for both of us.

Also, maybe there was some confusion earlier when you asked if I prefer these changes be made directly to rustfmt, and I responded with yes. In general, the team prefers changes to the subtree be made directly to the subtree. I wasn't trying to suggesting that you need to follow up with a PR on the rustfmt side for this. The rustfmt team could certainly handle making any necessary changes directly to rustfmt when we're ready.

I'm wondering if that clears things up and addresses your concern?

> I would also note, that clippy is also a subtree where I've made more significant changes in rust-lang/rust. Don't they have the same problem?

I'm not really familiar with the clippy team's perspective on these types of changes so It's hard to comment on that, but here's a [recent example where a PR author was asked to revert clippy changes related to dependency updates](https://github.com/rust-lang/rust/pull/120177#pullrequestreview-1835335290).

Reply:
> I'm wondering if that clears things up and addresses your concern?

Not really. Putting an `#[allow]` is the least disruptive change in terms of the code but not in terms of work to do, since either me or a rustdoc maintainer will at some point need to follow-up on this (since it's quite possibly on track for deny-by-default and/or hard error). While doing the change in this PR adds a bit of work when syncing the subtree, the change will be made (so no need to do any extra work, like remembering to do it, doing it, ...).

> I'm not really familiar with the clippy team's perspective on these types of changes so It's hard to comment on that, but here's a [recent example where a PR author was asked to revert clippy changes related to dependency updates](https://github.com/rust-lang/rust/pull/120177#pullrequestreview-1835335290).

I see it differently, updating a dependency (unless required of course) is a act regular development and as such should be done in "regular" way (so in clippy repo). The changes I had made (while not 100% required, since `#[allow]` also works) were made by necessity to have this PR pass, but also to have the least impact on the rustfmt team (since doing a dedicated PR takes time, can be long, ...).

-----

Anyway, I've changed my commit to only add an `#[allow]`.

Reply:
> Not really. Putting an #[allow] is the least disruptive change in terms of the code but not in terms of work to do, since either me or a rustdoc maintainer will at some point need to follow-up on this (since it's quite possibly on track for deny-by-default and/or hard error). While doing the change in this PR adds a bit of work when syncing the subtree, the change will be made (so no need to do any extra work, like remembering to do it, doing it, ...).

@Urgau Thank you for the change 🙏🏼 and for the follow up explanation. There's a lot of good info there that I wasn't aware of and wasn't considering. You're definitely more in the know on this RFC than I am. Do you know what the timeline looks like to transition this to deny-by-default and/or hard error? I'd be happy to take the lead and apply these changes on the rustfmt side when we get closer to that transition (or even before). I understand that this may mean a little extra work for me, but it's the workflow that the rustfmt team prefers right now.

Reply:
> Do you know what the timeline looks like to transition this to deny-by-default and/or hard error?

The RFC defines edition 2024 for a possible deny-by-default and 2027 (maybe) for a hard-error. With the allow you probably have until 2027 to move the code.

But I would advice moving it now since r-a doesn't handle `impl` definition inside bodies (functions, mod in functions, ...).